### PR TITLE
Core Schemas

### PR DESCRIPTION
**Pull request by CalebCourier**

### CODE REVIEW

File: schemas/core/reask.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+  "$id": "https://raw.githubusercontent.com/guardrails-ai/interfaces/core-schemas/schemas/core/reask.json",
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "ReAsk",
+  "type": "object",
+  "properties": {
+    "incorrectValue": {
```
Comment (zsimjee) [inline schemas/core/reask.json:7]: where do we keep the reaskPrompt? I don't see it in fail-result or validation-result. Is it an additionalPropery?

File: schemas/core/validation-outcome.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  "title": "ValidationOutcome",
+  "description": "The output from a Guard execution.",
+  "type": "object",
+  "properties": {
+    "rawLlmOutput": {
+      "type": "string",
+      "description": "The raw, unchanged string content from the LLM call."
+    },
+    "validatedOutput": {
+      "description": "The validated, and potentially fixed, output from the LLM call after undergoing validation.",
```
Comment (zsimjee) [inline schemas/core/validation-outcome.json:13]: I'm starting to dislike this definition XD

File: schemas/core/validator-log.json [hunk 3]
```diff
@@ -0,0 +1,57 @@
+{
+  "$id": "https://raw.githubusercontent.com/guardrails-ai/interfaces/core-schemas/schemas/core/validator-log.json",
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "ValidatorLog",
+  "type": "object",
+  "properties": {
+    "validatorName": {
+      "type": "string",
+      "description": "The class name of the validator.",
+      "format": "PascalCase"
```
Comment (zsimjee) [inline schemas/core/validator-log.json:10]: I always thought this was called TitleCase

File: schemas/core/validator-log.json [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    "propertyPath": {
+      "type": "string",
+      "description": "The JSON path to the property which was validated that produced this log."
+    },
+    "valueBeforeValidation": {
+      "$ref": "https://raw.githubusercontent.com/guardrails-ai/interfaces/core-schemas/schemas/core/any-type.json"
+    },
+    "valueAfterValidation": {
+      "$ref": "https://raw.githubusercontent.com/guardrails-ai/interfaces/core-schemas/schemas/core/any-type.json"
+    },
```
Comment (zsimjee) [inline schemas/core/validator-log.json:37]: I wish there were a way to lock valueAfterValidation to typeof(valueBeforeValidation) || null

**Review by zsimjee — COMMENTED** [review]
All other comments are either irrelevant or nits - 

Should we namespace some of the lower-level items? *-result, outputs, inputs, iteration, etc?

File: schemas/core/inputs.json [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{
+  "$id": "https://raw.githubusercontent.com/guardrails-ai/interfaces/core-schemas/schemas/core/inputs.json",
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "Inputs",
+  "type": "object",
+  "properties": {
+    "llmApi": {
```
[the remark this answers is not in the document]
Comment (CalebCourier) [reply schemas/core/inputs.json:7]: It already does actually!  So in python you set `Class.some_property`, but when it get's serialized to send to the api that becomes `Class.someProperty`

[the remark this answers is not in the document]
Comment (CalebCourier) [reply schemas/core/inputs.json:7]: It uses Pydantic aliases to do that.

File: schemas/core/inputs.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      "type": "string",
+      "description": "The LLM resource targeted by the user. e.g. openai.chat.completions.create"
+    },
+    "llmOutput": {
+      "type": "string",
+      "description": "The string output from an external LLM call provided by the user via Guard.parse."
```
[the remark this answers is not in the document]
Comment (CalebCourier) [reply schemas/core/inputs.json:13]: Like Polymorph it? If we keep `Guard.parse` around then yeah I think that's a good idea.  In general I imagined we would compress call and parse so it just becomes optional

File: schemas/core/outputs.json [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        "output": {
+          "type": "string"
+        },
+        "streamOutput": {
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (CalebCourier) [reply schemas/core/outputs.json:97]: probably, I can't remember if streaming gives you token count on each chunk or not...

File: schemas/core/reask.json (hunk 1 above)
Comment (CalebCourier) [reply schemas/core/reask.json:7]: It's currently stored on the schema, once these changes are implemented they go into an execution options container on the Guard.  Will probably become runtime/execution-time only options soon

File: schemas/core/validation-outcome.json (hunk 2 above)
Comment (CalebCourier) [reply schemas/core/validation-outcome.json:13]: same

File: schemas/core/validator-log.json (hunk 3 above)
Comment (CalebCourier) [reply schemas/core/validator-log.json:10]: technically title case is "Title Case"

File: schemas/core/validator-log.json (hunk 4 above)
Comment (CalebCourier) [reply schemas/core/validator-log.json:37]: I think with chaining these go away bc we'll no longer mutate values unless the user explicitly asks us to

**Comment by CalebCourier** [comment]
> All other comments are either irrelevant or nits -
> 
> Should we namespace some of the lower-level items? *-result, outputs, inputs, iteration, etc?

@zsimjee 
I generally avoid this since their specifics can be inferred from context; i.e. `Call.inputs`
That said, if we feel that the names could benefit from name-spacing it might just mean that they need better names in general.  I 100% believe this is the case for most of the History related schemas; I'm just not the right kind of creative to come up with something better that not also extremely verbose.

@nichwch and @wylansford open to suggestions if you have any insight

**Review by zsimjee — DISMISSED** [review]

**Review by zsimjee — APPROVED** [review]

**Pull request merged**